docs/closure-compiler.md

```diff
@@ -0,0 +1,697 @@
+1. https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System
+
+You can also check out Google Developers's [Getting Started with the UI](https://developers.google.com/closure/compiler/docs/gettingstarted_ui) tutorial which will
+introduce you to the [compiler appspot service](https://closure-compiler.appspot.com/home). The
+service is extremely useful for testing and debugging closure code in isolation.
+
+### Pro-tips for testing/debugging using the compiler appspot service
+
+If/when using the appspot service, make sure you have `ADVANCED_OPTIMIZATIONS`
+checked under the `optimization` option.
+
+Furthermore, you may want to check the `Pretty Print` option if you're debugging the output code.
+
+Additionally, if you use getter/setter properties within your code, add the following in between
+the `// ==ClosureCompiler==` / `// ==/ClosureCompiler==` comment block:
+
+```js
+// @language_out ECMASCRIPT5
+```
+
+This will tell closure to compile your code to ES5 (the default is ES3 which doesn't support
+accessor properties).
+
+Finally, you can make the permalinks to closure code a bit nicer by using a link shortener, such as
+https://goo.gl.
+
+You can use [this starter template](https://goo.gl/YSQkDi) to help debug your closure code, which
+has all of the above settings pre-configured.
+
+## MDC-Web Type System
+
+The following UML-like diagram shows a conceptual overview of the basic type system for MDC-Web. The
+diagram uses closure-esque type syntax, and represents what's in [mdc-base](../packages/mdc-base).
+
+![MDC-Web Type System UML(-like) diagram](https://docs.google.com/drawings/d/1mJBPiUkdFiXkU5A6kAdpZD5nXr6NkHIIW_vMNVdIvYY/pub?w=960&amp;h=720)
+
+> Note that the actual code to express this parameterization will vary slightly from the UML above,
+  since closure does not support bounded generics.
+
+The overall type system is relatively straightforward, and boils down to 3 main concepts:
+
+- **Adapters are simply [@record types](https://github.com/google/closure-compiler/wiki/Structural-Interfaces-in-Closure-Compiler) with a predefined shape specifying functions.** Because there
+  is no such thing as a "base adapter", they are simply considered to be plain objects.
+- **Foundations are parameterized by their adapters**. For example, an `MDCRippleFoundation` would be
+  parameterized by an `MDCRippleAdapter`. Thus, when declaring the `MDCRippleFoundation` class, the
+  proper JSDoc to specify this would be included: `@extends MDCFoundation<!MDCRippleAdapter>`.
+- **Components are parameterized by their foundations**. For example, an `MDCRipple` would be
+  parameterized by `MDCRippleFoundation`, which - as shown above - is itself parameterized by
+  `MDCRippleAdapter`. Thus, when declaring the `MDCRipple` class, the proper JSDoc to specify this
+  would be included: `@extends MDCComponent<!MDCRippleFoundation>`.
+
+## MDC-Web Closure Conventions
```

Comment:
This may be the single most useful section for ~~external~~ contributors. Consider making this more prominent somehow, or putting the link in the TL;DR